src/foam/util/concurrent/AsyncAssemblyLine.java

```diff
@@ -28,27 +28,39 @@ public AsyncAssemblyLine(X x) {
  public void enqueue(Assembly job) {
    final Assembly previous;
 
-   synchronized ( startLock_ ) {
-     previous = q_;
-     q_ = job;
-     job.startJob();
-   }
-
-   pool_.submit(x_, new ContextAgent() { public void execute(X x) {
-     job.executeJob();
-
-     if ( previous != null ) previous.waitToComplete();
-
-     synchronized ( endLock_ ) {
-       job.endJob();
-       job.complete();
+   try {
+     synchronized ( startLock_ ) {
+       previous = q_;
+       q_ = job;
+       job.executeUnderLock();
+       job.startJob();
      }
-
+   } catch (Throwable t) {
      // Isn't required, but helps GC last entry.
      synchronized ( startLock_ ) {
```

Comment:
I think this block should be moved outside of the catch, or put in and outside of the catch, or just removed altogether.